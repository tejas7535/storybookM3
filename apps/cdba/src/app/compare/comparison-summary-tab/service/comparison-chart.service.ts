import { Injectable } from '@angular/core';

import { TranslocoService } from '@jsverse/transloco';

@Injectable({ providedIn: 'root' })
export class ComparisonChartService {
  constructor(private readonly transloco: TranslocoService) {}

  readonly PURCHASE = this.transloco.translate(
    'compare.summary.common.purchase'
  );
  readonly REMAINDER = this.transloco.translate(
    'compare.summary.common.remainder'
  );
  readonly MOH = this.transloco.translate('compare.summary.common.moh');
  readonly BURDEN = this.transloco.translate('compare.summary.common.burden');
  readonly LABOUR = this.transloco.translate('compare.summary.common.labour');
  readonly RAW_MATERIAL = this.transloco.translate(
    'compare.summary.common.rawMaterial'
  );
  readonly TOTAL = this.transloco.translate('compare.summary.common.total');
}
