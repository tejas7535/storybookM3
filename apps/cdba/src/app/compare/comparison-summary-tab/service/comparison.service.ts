import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';

import { Observable } from 'rxjs';

import { API, ComparisonPath } from '@cdba/shared/constants/api';
import { BomIdentifier } from '@cdba/shared/models';

import { Comparison } from '../../../shared/models/comparison.model';

@Injectable({ providedIn: 'root' })
export class ComparisonService {
  constructor(private readonly httpClient: HttpClient) {}

  getComparison(identifiers: BomIdentifier[]): Observable<Comparison> {
    const path = `${API.v1}/${ComparisonPath}`;

    return this.httpClient.post<Comparison>(path, identifiers);
  }
}
