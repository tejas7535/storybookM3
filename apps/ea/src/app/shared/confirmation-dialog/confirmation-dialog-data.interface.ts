export interface ConfirmationDialogData {
  title: string;
  description: string;
  cancelActionText: string;
  confirmActionText: string;
}
