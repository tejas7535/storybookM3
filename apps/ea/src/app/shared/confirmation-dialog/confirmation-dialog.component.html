<div mat-dialog-title class="!px-0">
  <div
    class="inline-flex w-full flex-row content-around justify-between pl-4 pb-3 pr-2"
  >
    <div class="my-auto text-h6 text-high-emphasis">
      {{ data.title }}
    </div>
    <div class="flex flex-row">
      <button mat-icon-button (click)="closeDialog()">
        <mat-icon class="text-medium-emphasis">close</mat-icon>
      </button>
    </div>
  </div>
</div>

<mat-dialog-content class="!px-4 pb-4">
  <div>
    {{ data.description }}
  </div>
</mat-dialog-content>

<mat-dialog-actions class="justify-end">
  <button mat-stroked-button (click)="closeDialog()" color="primary">
    {{ data.cancelActionText }}
  </button>
  <button mat-raised-button color="primary" class="" (click)="applyDialog()">
    {{ data.confirmActionText }}
  </button>
</mat-dialog-actions>
