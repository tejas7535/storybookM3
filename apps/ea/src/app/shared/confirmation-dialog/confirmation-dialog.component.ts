import { Component, Inject } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { MatButtonModule } from '@angular/material/button';
import {
  MAT_DIALOG_DATA,
  MatDialogModule,
  MatDialogRef,
} from '@angular/material/dialog';
import { MatIconModule } from '@angular/material/icon';

import { ConfirmationDialogData } from './confirmation-dialog-data.interface';

@Component({
  imports: [
    MatDialogModule,
    MatIconModule,
    MatButtonModule,
    ReactiveFormsModule,
  ],
  templateUrl: './confirmation-dialog.component.html',
})
export class ConfirmationDialogComponent {
  constructor(
    public dialogRef: MatDialogRef<ConfirmationDialogComponent>,
    @Inject(MAT_DIALOG_DATA)
    public data: ConfirmationDialogData
  ) {}

  closeDialog(): void {
    this.dialogRef.close(false);
  }

  applyDialog(): void {
    this.dialogRef.close(true);
  }
}
