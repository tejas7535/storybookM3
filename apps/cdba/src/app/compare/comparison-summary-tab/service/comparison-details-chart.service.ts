import { Injectable } from '@angular/core';

import { TranslocoLocaleService } from '@jsverse/transloco-locale';
import { EChartsOption } from 'echarts';

import { DataPoint } from '@cdba/shared/components/bom-chart/data-point.model';
import { ComparisonType } from '@cdba/shared/constants';
import {
  CHART_HIGHLIGHT_ERROR,
  CHART_PRIMARY,
  CHART_SECONDARY,
  COST_SHARE_CATEGORY_COLORS,
} from '@cdba/shared/constants/colors';
import { ComparisonDetail } from '@cdba/shared/models/comparison.model';
import { CostShareService } from '@cdba/shared/services';

import { ComparisonChartService } from './comparison-chart.service';

@Injectable({ providedIn: 'root' })
export class ComparisonDetailsChartService {
  constructor(
    private readonly comparisonChartService: ComparisonChartService,
    private readonly localeService: TranslocoLocaleService,
    private readonly costShareService: CostShareService
  ) {}

  provideParetoChartConfig(
    details: ComparisonDetail[],
    currency: string
  ): EChartsOption {
    const xAxisData: string[] = [];
    const yAxisLineData: number[] = [];
    const yAxisBarDataPoints: DataPoint[] = [];
    let min = Number.POSITIVE_INFINITY;
    let max = Number.NEGATIVE_INFINITY;
    let cumulativePercentageCost = 0;

    details.forEach((detail) => {
      if (detail.costSharePercentage !== 0) {
        xAxisData.push(detail.title);
        min = Math.min(min, detail.totalCosts);
        max = Math.max(max, detail.totalCosts);

        cumulativePercentageCost += detail.costSharePercentage;
        cumulativePercentageCost = Number.parseFloat(
          cumulativePercentageCost.toPrecision(4)
        );
        yAxisLineData.push(cumulativePercentageCost);
        yAxisBarDataPoints.push({
          name: detail.title,
          value: detail.totalCosts,
          itemStyle: {
            color: COST_SHARE_CATEGORY_COLORS.get(
              this.costShareService.getCostShareCategory(
                detail.costSharePercentage / 100
              )
            ),
          },
        });
      }
    });

    return {
      tooltip: {
        trigger: 'axis',
        axisPointer: {
          type: 'cross',
          crossStyle: {
            color: '#999',
          },
        },
      },
      toolbox: {
        feature: {
          saveAsImage: { show: true },
        },
      },
      xAxis: [
        {
          type: 'category',
          data: xAxisData,
          axisPointer: {
            type: 'shadow',
          },
        },
      ],
      yAxis: [
        {
          type: 'value',
          min,
          max,
          axisLabel: {
            formatter: (value: string) => `${value} ${currency}`,
          },
        },
        {
          type: 'value',
          min: 0,
          max: 100,
          interval: 10,
          axisLabel: {
            formatter: (value: string) => `${value} %`,
          },
        },
      ],
      series: [
        {
          type: 'bar',
          tooltip: {
            valueFormatter: (value: number) =>
              `${this.localeService.localizeNumber(value, 'decimal')} ${currency}`,
          },
          data: yAxisBarDataPoints,
        },
        {
          type: 'line',
          color: CHART_HIGHLIGHT_ERROR,
          yAxisIndex: 1,
          tooltip: {
            valueFormatter: (value: string) => `${value} %`,
          },
          data: yAxisLineData,
        },
      ],
    } as EChartsOption;
  }

  provideDetailedBarChartConfig(
    detail: ComparisonDetail,
    firstMaterialDesignation: string,
    secondMaterialDesignation: string
  ): EChartsOption {
    const firstMaterialDesignationCosts: number[] = [];
    const secondMaterialDesignationCosts: number[] = [];

    detail.costDifferences.forEach((costDifference) => {
      switch (costDifference.type) {
        case ComparisonType.PURCHASE: {
          firstMaterialDesignationCosts[0] = costDifference.valueBom1;
          secondMaterialDesignationCosts[0] = costDifference.valueBom2;
          break;
        }
        case ComparisonType.RAW_MATERIAL: {
          firstMaterialDesignationCosts[1] = costDifference.valueBom1;
          secondMaterialDesignationCosts[1] = costDifference.valueBom2;
          break;
        }
        case ComparisonType.LABOUR: {
          firstMaterialDesignationCosts[2] = costDifference.valueBom1;
          secondMaterialDesignationCosts[2] = costDifference.valueBom2;
          break;
        }
        case ComparisonType.BURDEN: {
          firstMaterialDesignationCosts[3] = costDifference.valueBom1;
          secondMaterialDesignationCosts[3] = costDifference.valueBom2;
          break;
        }
        case ComparisonType.MOH: {
          firstMaterialDesignationCosts[4] = costDifference.valueBom1;
          secondMaterialDesignationCosts[4] = costDifference.valueBom2;
          break;
        }
        case ComparisonType.REMAINDER: {
          firstMaterialDesignationCosts[5] = costDifference.valueBom1;
          secondMaterialDesignationCosts[5] = costDifference.valueBom2;
          break;
        }
        case ComparisonType.TOTAL: {
          firstMaterialDesignationCosts[6] = costDifference.valueBom1;
          secondMaterialDesignationCosts[6] = costDifference.valueBom2;
          break;
        }
        default: {
          throw new Error(`Invalid Comparison Type: ${costDifference.type}`);
        }
      }
    });

    return {
      tooltip: {
        trigger: 'axis',
        axisPointer: {
          type: 'shadow',
        },
      },
      xAxis: {
        type: 'category',
        data: [
          this.comparisonChartService.PURCHASE,
          this.comparisonChartService.RAW_MATERIAL,
          this.comparisonChartService.LABOUR,
          this.comparisonChartService.BURDEN,
          this.comparisonChartService.MOH,
          this.comparisonChartService.REMAINDER,
          this.comparisonChartService.TOTAL,
        ],
        axisLabel: {
          interval: 0,
          width: 200,
        },
      },
      yAxis: {
        type: 'value',
      },
      series: [
        {
          type: 'bar',
          name: firstMaterialDesignation,
          data: firstMaterialDesignationCosts,
          itemStyle: {
            color: CHART_PRIMARY,
          },
        },
        {
          type: 'bar',
          name: secondMaterialDesignation,
          data: secondMaterialDesignationCosts,
          itemStyle: {
            color: CHART_SECONDARY,
          },
        },
      ],
    } as EChartsOption;
  }
}
