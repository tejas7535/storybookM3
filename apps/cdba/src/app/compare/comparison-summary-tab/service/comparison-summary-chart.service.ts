import { Injectable } from '@angular/core';

import { TranslocoService } from '@jsverse/transloco';
import { EChartsOption } from 'echarts/types/dist/echarts';

import {
  CHART_PRIMARY,
  CHART_SECONDARY,
  RADAR_CHART_REFERENCE_LINE,
} from '@cdba/shared/constants/colors';
import { ComparisonType } from '@cdba/shared/constants/comparison-type';
import {
  Comparison,
  ComparisonSummary,
  CostDifference,
} from '@cdba/shared/models/comparison.model';

import { ComparisonChartService } from './comparison-chart.service';

@Injectable({ providedIn: 'root' })
export class ComparisonSummaryChartService {
  constructor(
    private readonly transloco: TranslocoService,
    private readonly comparisonChartService: ComparisonChartService
  ) {}

  provideSummaryTableColumns(): string[] {
    return [
      'costType',
      'valueBom1',
      'valueBom2',
      'currency',
      'costDeltaValue',
      'costDeltaPercentage',
    ];
  }

  provideSummarizedRadarChartConfig(
    firstMaterialDesignation: string,
    secondMaterialDesignation: string,
    comparison: Comparison
  ): EChartsOption {
    const referenceLineTitle = this.transloco.translate(
      'compare.summary.common.referenceLine'
    );

    const purchaseCost = this.findCostDifferenceByComparisonType(
      comparison.summary,
      ComparisonType.PURCHASE
    );
    const remainderCost = this.findCostDifferenceByComparisonType(
      comparison.summary,
      ComparisonType.REMAINDER
    );
    const mohCost = this.findCostDifferenceByComparisonType(
      comparison.summary,
      ComparisonType.MOH
    );
    const burdenCost = this.findCostDifferenceByComparisonType(
      comparison.summary,
      ComparisonType.BURDEN
    );
    const labourCost = this.findCostDifferenceByComparisonType(
      comparison.summary,
      ComparisonType.LABOUR
    );
    const rawMaterialCost = this.findCostDifferenceByComparisonType(
      comparison.summary,
      ComparisonType.RAW_MATERIAL
    );

    // Calculate min max values
    let min = Number.POSITIVE_INFINITY;
    let max = Number.NEGATIVE_INFINITY;
    comparison.summary.costDifferences.forEach((costDifference) => {
      if (costDifference.type !== 'TOTAL') {
        min = Math.min(min, costDifference.valueBom1, costDifference.valueBom2);
        max = Math.max(max, costDifference.valueBom1, costDifference.valueBom2);
      }
    });
    min = Math.floor(min);
    max = Math.ceil(max);

    return {
      legend: {
        bottom: '0%',
        left: '5%',
        data: [
          referenceLineTitle,
          firstMaterialDesignation,
          secondMaterialDesignation,
        ],
      },
      tooltip: {
        trigger: 'item',
        valueFormatter: (value: string) => `${value} ${comparison.currency}`,
      },
      toolbox: {
        feature: {
          saveAsImage: { show: true },
        },
      },
      radar: {
        indicator: [
          {
            name: this.comparisonChartService.PURCHASE,
            min,
            max,
          },
          {
            name: this.comparisonChartService.REMAINDER,
            min,
            max,
          },
          {
            name: this.comparisonChartService.MOH,
            min,
            max,
          },
          {
            name: this.comparisonChartService.BURDEN,
            min,
            max,
          },
          {
            name: this.comparisonChartService.LABOUR,
            min,
            max,
          },
          {
            name: this.comparisonChartService.RAW_MATERIAL,
            min,
            max,
          },
        ],
      },
      series: [
        {
          type: 'radar',
          data: [
            {
              name: referenceLineTitle,
              value: [0, 0, 0, 0, 0, 0],
              itemStyle: {
                color: RADAR_CHART_REFERENCE_LINE,
              },
            },
            {
              name: firstMaterialDesignation,
              value: [
                purchaseCost.valueBom1,
                remainderCost.valueBom1,
                mohCost.valueBom1,
                burdenCost.valueBom1,
                labourCost.valueBom1,
                rawMaterialCost.valueBom1,
              ],
              itemStyle: {
                color: CHART_PRIMARY,
              },
            },
            {
              name: secondMaterialDesignation,
              value: [
                purchaseCost.valueBom2,
                remainderCost.valueBom2,
                mohCost.valueBom2,
                burdenCost.valueBom2,
                labourCost.valueBom2,
                rawMaterialCost.valueBom2,
              ],
              itemStyle: {
                color: CHART_SECONDARY,
              },
            },
          ],
        },
      ],
    } as EChartsOption;
  }

  findCostDifferenceByComparisonType(
    summary: ComparisonSummary,
    type: ComparisonType
  ): CostDifference {
    return summary.costDifferences.find(
      (costDifference) => costDifference.type === type
    );
  }
}
