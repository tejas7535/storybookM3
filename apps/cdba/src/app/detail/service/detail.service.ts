import { HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';

import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';

import { DataService } from '../../core/http/data.service';
import {
  BomIdentifier,
  BomItem,
  BomResult,
  ReferenceTypeIdModel,
  ReferenceTypeResultModel,
} from '../../core/store/reducers/detail/models';
import { CalculationsResultModel } from '../../core/store/reducers/detail/models/calculations-result-model';
import { Calculation } from '../../core/store/reducers/shared/models/calculation.model';

@Injectable({
  providedIn: 'root',
})
export class DetailService {
  private readonly DETAIL = 'detail';

  private readonly CALCULATIONS = 'calculations';

  private readonly PARAM_MATERIAL_NUMBER = 'material-number';
  private readonly PARAM_PLANT = 'plant';

  private readonly BOM = 'bom';
  private readonly PARAM_BOM_COSTING_DATE = 'bom-costing-date';
  private readonly PARAM_BOM_COSTING_NUMBER = 'bom-costing-number';
  private readonly PARAM_BOM_COSTING_TYPE = 'bom-costing-type';
  private readonly PARAM_BOM_COSTING_VERSION = 'bom-costing-version';
  private readonly PARAM_BOM_ENTERED_MANUALLY = 'bom-entered-manually';
  private readonly PARAM_BOM_REFERENCE_OBJECT = 'bom-reference-object';
  private readonly PARAM_BOM_VALUATION_VARIANT = 'bom-valuation-variant';

  private static defineBomTreeForAgGrid(
    items: BomItem[],
    idx: number
  ): BomItem[] {
    if (idx === items.length) {
      return items;
    }

    const currentItem = items[idx];
    const nextIdx = idx + 1;

    // special case level 1
    if (currentItem.level === 1) {
      currentItem.predecessorsInTree = [currentItem.materialDesignation];

      return DetailService.defineBomTreeForAgGrid(items, nextIdx);
    }

    const previousItem = items[idx - 1];

    if (previousItem.level < currentItem.level) {
      // previous item is parent
      currentItem.predecessorsInTree = [
        ...previousItem.predecessorsInTree,
        currentItem.materialDesignation,
      ];
    } else {
      // previous item has same parent or different parent
      const itemsToRemove = previousItem.level - currentItem.level + 1;
      const predecessors = [...previousItem.predecessorsInTree];

      // remove last items to have the actual parent as last item
      predecessors.splice(predecessors.length - itemsToRemove, itemsToRemove);

      // since duplicates are possible, add spaces to the key until it is unique
      let foundUniqueKey = false;
      let currentMaterialDesignation = currentItem.materialDesignation;
      while (!foundUniqueKey) {
        const temp = [...predecessors, currentMaterialDesignation];

        if (
          items
            .map((item) => item.predecessorsInTree)
            .find(
              (materialDesignations: string[]) =>
                materialDesignations.length === temp.length &&
                materialDesignations.every(
                  (value: string, index: number) => value === temp[index]
                )
            ) !== undefined
        ) {
          currentMaterialDesignation += ' ';
        } else {
          foundUniqueKey = true;
        }
      }

      currentItem.predecessorsInTree = [
        ...predecessors,
        currentMaterialDesignation,
      ];
    }

    return DetailService.defineBomTreeForAgGrid(items, nextIdx);
  }

  public constructor(private readonly dataService: DataService) {}

  public detail(
    item: ReferenceTypeIdModel
  ): Observable<ReferenceTypeResultModel> {
    const params: HttpParams = new HttpParams()
      .set(this.PARAM_MATERIAL_NUMBER, item.materialNumber)
      .set(this.PARAM_PLANT, item.plant);

    return this.dataService.getAll<ReferenceTypeResultModel>(
      this.DETAIL,
      params
    );
  }

  public calculations(materialNumber: string): Observable<Calculation[]> {
    const params = new HttpParams().set(
      this.PARAM_MATERIAL_NUMBER,
      materialNumber
    );

    return this.dataService
      .getAll<CalculationsResultModel>(this.CALCULATIONS, params)
      .pipe(map((response: CalculationsResultModel) => response.items));
  }

  public getBom(bomIdentifier: BomIdentifier): Observable<BomItem[]> {
    const params = new HttpParams()
      .set(this.PARAM_BOM_COSTING_DATE, bomIdentifier.bomCostingDate)
      .set(this.PARAM_BOM_COSTING_NUMBER, bomIdentifier.bomCostingNumber)
      .set(this.PARAM_BOM_COSTING_TYPE, bomIdentifier.bomCostingType)
      .set(this.PARAM_BOM_COSTING_VERSION, bomIdentifier.bomCostingVersion)
      .set(this.PARAM_BOM_ENTERED_MANUALLY, bomIdentifier.bomEnteredManually)
      .set(this.PARAM_BOM_REFERENCE_OBJECT, bomIdentifier.bomReferenceObject)
      .set(this.PARAM_BOM_VALUATION_VARIANT, bomIdentifier.bomValuationVariant);

    return this.dataService.getAll<BomResult>(this.BOM, params).pipe(
      map((response: BomResult) =>
        response.items.map((item) => ({
          ...item,
          predecessorsInTree: [],
        }))
      ),
      map((items: BomItem[]) => DetailService.defineBomTreeForAgGrid(items, 0))
    );
  }
}
