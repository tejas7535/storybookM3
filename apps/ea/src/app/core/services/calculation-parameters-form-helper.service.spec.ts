/* eslint-disable unicorn/no-null */
import { FormGroup } from '@angular/forms';
import {
  MatDialog,
  MatDialogModule,
  MatDialogRef,
} from '@angular/material/dialog';

import { LoadCaseDataFormGroupModel } from '@ea/calculation/calculation-parameters/loadcase-data-form-group.interface';
import { ConfirmationDialogComponent } from '@ea/shared/confirmation-dialog/confirmation-dialog.component';
import { TranslocoService } from '@jsverse/transloco';
import { createServiceFactory, SpectatorService } from '@ngneat/spectator/jest';

import { provideTranslocoTestingModule } from '@schaeffler/transloco/testing';

import {
  CalculationParametersOperationConditions,
  LoadCaseData,
} from '../store/models';
import { CalculationParametersFormHelperService } from './calculation-parameters-form-helper.service';

describe('CalculationParametersFormHelperService', () => {
  let service: CalculationParametersFormHelperService;
  let spectator: SpectatorService<CalculationParametersFormHelperService>;
  let translocoService: TranslocoService;
  let translocoServiceSpy: jest.SpyInstance;

  const createService = createServiceFactory({
    service: CalculationParametersFormHelperService,
    imports: [MatDialogModule, provideTranslocoTestingModule({ en: {} })],
    providers: [
      CalculationParametersFormHelperService,
      {
        provide: MatDialog,
        useValue: {
          open: jest.fn(() => ({}) as unknown as MatDialogRef<any>),
        },
      },
    ],
  });

  const createMockLoadCaseWithOperatingTime = (
    operatingTimeValue: number | undefined
  ): FormGroup<LoadCaseDataFormGroupModel> =>
    ({
      controls: {
        operatingTime: {
          value: operatingTimeValue,
        },
      },
    }) as unknown as FormGroup<LoadCaseDataFormGroupModel>;

  beforeEach(() => {
    spectator = createService();
    service = spectator.service;
    translocoService = spectator.inject(TranslocoService);
    translocoServiceSpy = jest.spyOn(translocoService, 'translate');
  });

  it('should be created', () => {
    expect(service).toBeDefined();
  });

  describe('getTotalOperatingTimeForLoadcases', () => {
    it('should provide value for empty loadcases', () => {
      const loadCases: FormGroup<LoadCaseDataFormGroupModel>[] = [];

      const result = service.getTotalOperatingTimeForLoadcases(loadCases);

      expect(result).toBe(0);
    });

    it('should provide total value for few loadcases', () => {
      const loadCases: FormGroup<LoadCaseDataFormGroupModel>[] = [
        createMockLoadCaseWithOperatingTime(20),
        createMockLoadCaseWithOperatingTime(50),
        createMockLoadCaseWithOperatingTime(undefined),
        createMockLoadCaseWithOperatingTime(25),
        createMockLoadCaseWithOperatingTime(25),
      ];

      const result = service.getTotalOperatingTimeForLoadcases(loadCases);
      expect(result).toBe(120);
    });
  });

  describe('getLocalizedLoadCaseName', () => {
    it('should get localized load case name', () => {
      const result = service.getLocalizedLoadCaseName(2);

      expect(translocoServiceSpy).toHaveBeenCalledWith(
        'operationConditions.loadCaseName',
        {
          number: 2,
        }
      );

      expect('operationConditions.loadCaseName').toBe(result);
    });
  });

  describe('openConfirmDeleteDialog', () => {
    it('should open the dialog', () => {
      const mockRef = service.openConfirmDeleteDialog();

      expect(mockRef).toEqual({} as unknown as MatDialogRef<any>);
      expect(service['dialog'].open).toHaveBeenCalledWith(
        ConfirmationDialogComponent,
        {
          data: {
            cancelActionText:
              'operationConditions.confirmationDialog.cancelAction',
            confirmActionText:
              'operationConditions.confirmationDialog.confirmationAction',
            description: 'operationConditions.confirmationDialog.description',
            title: 'operationConditions.confirmationDialog.title',
          },
          width: '500px',
          autoFocus: false,
        }
      );
    });
  });

  describe('getLocalizedLoadCaseTimePortion', () => {
    describe('when operatingTimeInHours and loadcasePercentage are not provided', () => {
      it('should return empty string', () => {
        const result = service.getLocalizedLoadCaseTimePortion(
          undefined,
          undefined
        );

        expect(result).toBe('');
      });
    });

    describe('when operatingTimeInHours and loadcasePercentage are provided', () => {
      it('should return localized load case time portion', () => {
        const operatingTimeInHours = 100;
        const loadcasePercentage = 25;
        const expectedLoadcaseTime = 25;
        const expectedTranslation = 'operationConditions.loadCaseTimePortion';

        const result = spectator.service.getLocalizedLoadCaseTimePortion(
          operatingTimeInHours,
          loadcasePercentage
        );

        expect(translocoServiceSpy).toHaveBeenCalledWith(
          'operationConditions.loadCaseTimePortion',
          {
            loadcasePercentage,
            totalTime: operatingTimeInHours,
            loadcaseTime: expectedLoadcaseTime,
          }
        );
        expect(result).toBe(expectedTranslation);
      });
    });

    describe('when updating result to handle negative values', () => {
      it('should replace null values with undefined', () => {
        const previousInput: Partial<CalculationParametersOperationConditions> =
          {
            ambientTemperature: undefined,
            time: undefined,
            loadCaseData: [{ operatingTemperature: undefined } as LoadCaseData],
          };

        const newInput: Partial<CalculationParametersOperationConditions> = {
          ambientTemperature: null,
          time: null,
          loadCaseData: [{ operatingTemperature: null } as LoadCaseData],
        };

        const result = service.updateResultsToHandleNegativeValues(
          previousInput,
          newInput
        );
        expect(result).toMatchSnapshot();
      });

      it('should not change non-null values', () => {
        const previousInput = {
          ambientTemperature: 25,
          time: 123,
          loadCaseData: [{ operatingTemperature: 20 } as LoadCaseData],
        };

        const newInput = {
          ambientTemperature: 25,
          time: 123,
          loadCaseData: [{ operatingTemperature: 20 } as LoadCaseData],
        };

        const result = service.updateResultsToHandleNegativeValues(
          previousInput,
          newInput
        );
        expect(result).toMatchSnapshot();
      });

      describe('when new input is null with minus sign scenario', () => {
        it('should set previous value to undefined to block change detection', () => {
          const previousInput = {
            ambientTemperature: 25,
            time: 123,
            loadCaseData: [{ operatingTemperature: 20 } as LoadCaseData],
          };

          const newInput: Partial<CalculationParametersOperationConditions> = {
            ambientTemperature: null,
            time: 123,
            loadCaseData: [{ operatingTemperature: null } as LoadCaseData],
          };

          const result = service.updateResultsToHandleNegativeValues(
            previousInput,
            newInput
          );
          expect(result).toMatchSnapshot();
        });
      });
    });
  });
});
