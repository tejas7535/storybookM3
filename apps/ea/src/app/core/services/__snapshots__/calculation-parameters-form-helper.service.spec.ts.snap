// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalculationParametersFormHelperService getLocalizedLoadCaseTimePortion when updating result to handle negative values should not change non-null values 1`] = `
Object {
  "new": Object {
    "ambientTemperature": 25,
    "loadCaseData": Array [
      Object {
        "operatingTemperature": 20,
      },
    ],
    "time": 123,
  },
  "previous": Object {
    "ambientTemperature": 25,
    "loadCaseData": Array [
      Object {
        "operatingTemperature": 20,
      },
    ],
    "time": 123,
  },
}
`;

exports[`CalculationParametersFormHelperService getLocalizedLoadCaseTimePortion when updating result to handle negative values should replace null values with undefined 1`] = `
Object {
  "new": Object {
    "ambientTemperature": undefined,
    "loadCaseData": Array [
      Object {
        "operatingTemperature": undefined,
      },
    ],
    "time": null,
  },
  "previous": Object {
    "ambientTemperature": undefined,
    "loadCaseData": Array [
      Object {
        "operatingTemperature": undefined,
      },
    ],
    "time": undefined,
  },
}
`;

exports[`CalculationParametersFormHelperService getLocalizedLoadCaseTimePortion when updating result to handle negative values when new input is null with minus sign scenario should set previous value to undefined to block change detection 1`] = `
Object {
  "new": Object {
    "ambientTemperature": undefined,
    "loadCaseData": Array [
      Object {
        "operatingTemperature": undefined,
      },
    ],
    "time": 123,
  },
  "previous": Object {
    "ambientTemperature": undefined,
    "loadCaseData": Array [
      Object {
        "operatingTemperature": undefined,
      },
    ],
    "time": 123,
  },
}
`;
